// @lynx-js/react/runtime/lib/lifecycle/patch/snapshotPatch.js
// Copyright 2024 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.
export var SnapshotOperation;
(function (SnapshotOperation) {
    SnapshotOperation[SnapshotOperation["CreateElement"] = 0] = "CreateElement";
    SnapshotOperation[SnapshotOperation["InsertBefore"] = 1] = "InsertBefore";
    SnapshotOperation[SnapshotOperation["RemoveChild"] = 2] = "RemoveChild";
    SnapshotOperation[SnapshotOperation["SetAttribute"] = 3] = "SetAttribute";
    SnapshotOperation[SnapshotOperation["SetAttributes"] = 4] = "SetAttributes";
    SnapshotOperation[SnapshotOperation["DEV_ONLY_AddSnapshot"] = 100] = "DEV_ONLY_AddSnapshot";
    SnapshotOperation[SnapshotOperation["DEV_ONLY_RegisterWorklet"] = 101] = "DEV_ONLY_RegisterWorklet";
})(SnapshotOperation || (SnapshotOperation = {}));
export let __globalSnapshotPatch;
export function takeGlobalSnapshotPatch() {
    if (__globalSnapshotPatch) {
        const list = __globalSnapshotPatch;
        __globalSnapshotPatch = [];
        return list;
    }
    else {
        return undefined;
    }
}
export function initGlobalSnapshotPatch() {
    __globalSnapshotPatch = [];
}
export function deinitGlobalSnapshotPatch() {
    __globalSnapshotPatch = undefined;
}
